### PR TITLE
Decorators, closes #137

### LINKED ISSUE #137: Document decorators

### PR DESCRIPTION
**Pull request by familyfriendlymikey**

### CODE REVIEW

**Comment by trafnar** [comment]
I'm glad to have this documented, this is all new to me :) how did you even figure all this out?!

My first point of feedback: you switch from showing `def calc` as a normal function, to having it be a method on a class when you introduce the decorator to it, but there is no explanation of that. Do these only work in classes? Maybe the initial example should be for a class method as well.

**Comment by trafnar** [comment]
I also think this needs a more clear explanation of why you would want a decorator at all - the logging example works, but I'm left wondering "so in what cases would I actually do this?" I think we can point to some of the built-in decorators (like `@autorun` or whatever?) as examples?

**Comment by trafnar** [comment]
(I'm making notes as I go through this, and I'm learning this feature as I go)

Let's maybe use the same argument names as the MDN docs for consistency?

`def @example obj, prop, descriptor` instead of `def @example target, name, desc`

**Comment by trafnar** [comment]
> You can define and use decorators to replace a property with a new descriptor at runtime.

Is a decorator necessary to do that? Can you do it with normal JS? Is a decorator just a more convenient way to do it?

**Comment by trafnar** [comment]
in the first section:
> See also [additional details](http://localhost:3000/docs/decorators#additional-details).

seems unnecessary or unclear, why do I want additional details instead of reading the next section?

**Comment by trafnar** [comment]
I pushed an update with some tweaks to the language to make it a bit more clear, based on my experience learning from this guide.

**Comment by trafnar** [comment]
I think having this page is much better than having nothing, and there's nothing wrong with it, so we should merge it, but I really think it just needs some explanation of what the use cases are, why you would use this, other than logging.

**Comment by familyfriendlymikey** [comment]
Thanks for giving this a look :) Definitely took a while, had to dig through Imba's code base to figure it out.

Questions/responses for you here, everything else I agree with and will edit accordingly.

> I'm left wondering  "so in what cases would I actually do this?"

I completely agree, that was my main struggle with writing this since I really dislike fake examples. I kind of ended up deciding that this was just a more advanced use case. But on that note, I remember discussing a function like this with you:

```
def handle_click_config
	const doStuff = do
		...

	loading = yes
	doStuff()
	settings_active = no
	loading = no
```

This pattern appears a lot in my apps, where I set/unset something before/after a function, and if the function returns early it still needs to do the unsetting part, which is why I put the actual logic inside another nested function.

I realized that this can just be implemented with decorators:

```
@load def handle_click_config
	...
```

But I wasn't sure if this pattern was something other people use in their code so I just opted for the whole `log` thing since it was much simpler to explain.

If you think it's a good example I can add it to the guide, or if any better ones come to mind let me know. Otherwise I'll just leave it as `log` for now.

> I think we can point to some of the built-in decorators (like @autorun or whatever?) as examples?

Was planning on having `observable` being a sub-menu of decorators in the nav. Do you think that suffices?

> Let's maybe use the same argument names as the MDN docs for consistency?

I don't mind either way, I was just using what Sindre had in the Imba codebase. Maybe he had a reason for it, I'll leave this for now and slip it by him on a call or something.

> seems unnecessary or unclear, why do I want additional details instead of reading the next section?

I'm trying to balance between giving terse, essential information to experienced users and a ground-up holistic explanation to beginners on the same page, which is why I did the whole TLDR -> Additional Details thing.

The TLDR is just so a reader can see what the page is about. The intended flow is that if they're a beginner, they'll naturally continue onto the guide, but if they're an experienced user, they'll avoid the guide and click the link to additional details to find what they're looking for.

To your point, maybe the issue is the wording of "additional details", which I can see now isn't ideal. I still think the layout is interesting -- having only an entire guide would be pretty annoying to experienced users imo since they'd have to dig harder for information. If you have other ideas though I'd be interested.

**Comment by trafnar** [comment]
One thing that would go a long way is just a better description. The first sentence:

> Decorators can be used to replace a property with a new descriptor at runtime.

Should describe it better. What is the problem to solve? I've never felt "I need to change this descriptor at runtime" but I know I have had problems that could be solved by decorators. 

I'm trying to think of something better, but it's tough, and I don't have an answer yet :)

Something to the effect "Decorators are for wrapping properties and methods with additional code to extend their functionality"...

**Comment by trafnar** [comment]
> Was planning on having observable being a sub-menu of decorators in the nav. Do you think that suffices?

I think it's worth a mention/link like "here is one of the main ways imba uses this"

**Comment by trafnar** [comment]
> I'm trying to balance between giving terse, essential information to experienced users and a ground-up holistic explanation to beginners on the same page, which is why I did the whole TLDR -> Additional Details thing.

Ok, I think you're totally right about your approach to the structure of the page, but like you said, we can word it differently.

How about something like "Jump to" instead of "See also", and we call it "Technical Details" instead of "Additional Information"

Maybe also "Guide" could be called something like "Guide to Understanding" 

Also the 'see also additional details' link doesn't seem to work for me.

**Comment by trafnar** [comment]
I made a bunch of updates in b4dbbd6, cut out some parts, tried to shorten everything, hope you don't mind :)

I tried to make the example more realistic by calling it a debugging tool rather than just logging for no real reason. Summary:

> You later realize you want to log the result of that calculation for debugging purposes, now you need to assign the calculation to a variable, then you can log the value, then return it.
> We can make a convenient, reusable decorator that automatically logs the return value, instead of having to modify the function with additional code.
> Here's how we can define such a decorator...

I made some code changes to be more explicit in some parts (like using `return` in some places) just to make it extra clear for learners. I also used `()` on function calls in some places to help demonstrate that both are usable.

I included arguments in the example function `multiply a, b do a * b` instead of `multiply do 5 * 5` I think it doesn't add too much complexity and allows us to cover all the details with less explanation/code.

I also put everything into classes/methods instead of regular functions at the start because I believe that decorators are only for properties of classes.

Couple outstanding things:
- [ ] I mention `@observable` at the beginning, that should link to the observable page, but that page is in a different branch right now
- [ ] In the 'return value' section, I don't quite understand how returning 'null' is supposed to work, the example doesn't work the way I expected.
- [ ] I deleted the higher order function example, I think it's too much explanation and we can get the information across without it. However, that example is referenced later in the 'return value' section, but I'm a little mixed up about that section so I haven't changed those references yet (until I better understand how the return value works).

**Comment by trafnar** [comment]
I've made some modifications, I think this is in a usable state now.

**Comment by familyfriendlymikey** [comment]
Thanks for reviewing this, sorry it took me so long to get to it!
